sdk/cosmosdb/cosmos/src/bulk/BulkStreamerPerPartition.ts

```diff
@@ -0,0 +1,145 @@
+import type { ItemBulkOperation } from "./ItemBulkOperation";
+import type { DiagnosticNodeInternal } from "../diagnostics/DiagnosticNodeInternal";
+import type { RequestOptions } from "../request/RequestOptions";
+import type { BulkOperationResult } from "./BulkOperationResult";
+import { BulkPartitionMetric } from "./BulkPartitionMetric";
+import { BulkCongestionAlgorithm } from "./BulkCongestionAlgorithm";
+import type { Limiter } from "./Limiter";
+
+/**
+ * Handles operation queueing and dispatching. Fills batches efficiently and maintains a timer for early dispatching in case of partially-filled batches and to optimize for throughput.
+ * There is always one batch at a time being filled. Locking is in place to avoid concurrent threads trying to Add operations while the timer might be Dispatching the current batch.
+ * The current batch is dispatched and a new one is readied to be filled by new operations, the dispatched batch runs independently through a fire and forget pattern.
+ * @hidden
+ */
+
+export class BulkStreamerPerPartition {
+  private readonly executor: ExecuteCallback;
+  private readonly retrier: RetryCallback;
+  private readonly options: RequestOptions;
+  private readonly diagnosticNode: DiagnosticNodeInternal;
+
+  private currentBatcher: BulkBatcher;
+  private readonly lock: semaphore.Semaphore;
+  private dispatchTimer: NodeJS.Timeout;
+  private readonly orderedResponse: BulkOperationResult[] = [];
+  private limiterSemaphore: Limiter;
+
+  private readonly oldPartitionMetric: BulkPartitionMetric;
+  private readonly partitionMetric: BulkPartitionMetric;
+  private congestionDegreeOfConcurrency = 1;
+  private congestionControlAlgorithm: BulkCongestionAlgorithm;
+
+  constructor(
+    executor: ExecuteCallback,
+    retrier: RetryCallback,
+    limiter: Limiter,
+    options: RequestOptions,
+    diagnosticNode: DiagnosticNodeInternal,
+    orderedResponse: BulkOperationResult[],
+  ) {
+    this.executor = executor;
+    this.retrier = retrier;
+    this.limiterSemaphore = limiter;
+    this.options = options;
+    this.diagnosticNode = diagnosticNode;
+    this.orderedResponse = orderedResponse;
+    this.oldPartitionMetric = new BulkPartitionMetric();
+    this.partitionMetric = new BulkPartitionMetric();
+    this.congestionControlAlgorithm = new BulkCongestionAlgorithm(
+      this.limiterSemaphore,
+      this.partitionMetric,
+      this.oldPartitionMetric,
+    );
+    this.currentBatcher = this.createBulkBatcher();
+
+    this.lock = semaphore(1);
+    this.runDispatchTimer();
+  }
+
+  /**
+   * adds a bulk operation to current batcher and dispatches if batch is full
+   * @param operation - operation to add
+   */
+  add(operation: ItemBulkOperation): void {
+    let toDispatch: BulkBatcher | null = null;
+    this.lock.take(() => {
+      try {
+        // attempt to add operation until it fits in the current batch for the streamer
+        while (!this.currentBatcher.tryAdd(operation)) {
```

Comment:
If there's a backend issue where the backend keeps timing out and, after some retries, returns an error, the dispatcher might get stuck. In such cases, does it really make sense to keep a thread busy while adding the batcher? Wouldn't it be better to make this method async and wait for `tryAdd` to complete? Maybe the batcher can handle the issue internally.